File: ios/Classes/FlutterChacha20Poly1305Plugin.swift
import Flutter
import UIKit
import CryptoKit

enum CryptoError: Error {
    case runtimeError(String)
}

@available(iOS 13.0, *)
public class FlutterChacha20Poly1305Plugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_chacha20_poly1305", binaryMessenger: registrar.messenger())
    let instance = FlutterChacha20Poly1305Plugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      guard let args = call.arguments as? [String: Any] else {
          result(FlutterError(
              code: "CAUGHT_ERROR",
              message:"Invalid arguments",
              details: nil))
          return
      }
      
      do {
        switch call.method {
        case "encrypt":
            try self.encrypt(args: args, result:result)
        case "encryptString":
            try self.encryptString(args: args, result:result)
        case "decrypt":
            try self.decrypt(args: args, result:result)
        case "decryptString":
            try self.decryptString(args: args, result:result)
        default:
            result(FlutterMethodNotImplemented)
        }
      } catch let error as NSError {
          result(FlutterError(
              code: "CAUGHT_ERROR",
              message:"\(error.domain), \(error.code), \(error.description)",
              details: nil))
      } catch {
          result(FlutterError(
              code: "CAUGHT_ERROR",
              message:"\(error)",
              details: nil))
      }
  }
    
    private func encryptData(plainData: Data, key: Data) throws -> ChaChaPoly.SealedBox {
        let skey = SymmetricKey(data: key)
        return try ChaChaPoly.seal(plainData, using: skey)
    }

    private func decryptData(cipherData: Data, keyData: Data, nonceData: Data, tagData: Data) throws -> Data {
        let skey = SymmetricKey(data: keyData)
        let sealedBox = try ChaChaPoly.SealedBox(nonce: ChaChaPoly.Nonce(data: nonceData),
                                               ciphertext: cipherData,
                                               tag: tagData)
        let decryptedData = try ChaChaPoly.open(sealedBox, using: skey)
        return decryptedData
    }

    private func encrypt(args: [String: Any], result: @escaping FlutterResult) throws {
        do {
            guard let data = (args["data"] as? FlutterStandardTypedData)?.data else {
                result(parameterError(name: "data"))
                return
            }
            guard let key = (args["key"] as? FlutterStandardTypedData)?.data else {
                result(parameterError(name: "key"))
                return
            }

            let sealed = try self.encryptData(plainData: data, key: key)

            result([
                "encrypted": FlutterStandardTypedData(bytes: sealed.ciphertext),
                "tag": FlutterStandardTypedData(bytes: sealed.tag),
                "nonce": FlutterStandardTypedData(bytes: sealed.nonce.withUnsafeBytes {
                    Data(Array($0))
                }),
            ])
        } catch CryptoError.runtimeError(let errorMessage) {
            result(FlutterError(code: "InvalidArgumentError", message: errorMessage.localizedLowercase, details: nil))
        } catch {
            result(FlutterError(code: "EncryptionError", message:"Failed to encrypt: " + error.localizedDescription, details: nil))
        }
    }
    
    private func encryptString(args: [String: Any], result: @escaping FlutterResult) throws {
        do {
            guard let keyEncoding = args["keyEncoding"] as? String else {
                result(parameterError(name: "keyEncoding"))
                return
            }

            guard let outputEncoding = args["outputEncoding"] as? String else {
                result(parameterError(name: "outputEncoding"))
                return
            }
            
            
            if (keyEncoding != "base64" && keyEncoding != "hex") {
                result(FlutterError(code: "Input encoding eror", message: "Input key encoding should be in 'base64' or 'hex'", details: nil))
                return
            }

            if (outputEncoding != "base64" && outputEncoding != "hex") {
                result(FlutterError(code: "Output encoding eror", message: "Output encoding should be in 'base64' or 'hex'", details: nil))
                return
            }

            guard let key = args["key"] as? String else {
                result(parameterError(name: "key"))
                return
            }
            
            guard let string = args["string"] as? String else {
                result(parameterError(name: "string"))
                return
            }
            
            let keyData = keyEncoding == "base64" ? Data(base64Encoded: key)! : Data(hexString: key)!
            let plainData = string.data(using: .utf8)!

            let sealed = try self.encryptData(plainData: plainData, key: keyData)
            
            let encrypted = outputEncoding == "base64" ? sealed.ciphertext.base64EncodedString() : sealed.ciphertext.hexadecimal
            let nonce = outputEncoding == "base64" ? sealed.nonce.withUnsafeBytes {
                Data(Array($0)).base64EncodedString()
            } : sealed.nonce.withUnsafeBytes {
                Data(Array($0)).hexadecimal
            }
            let tag = outputEncoding == "base64" ? sealed.tag.base64EncodedString() : sealed.tag.hexadecimal

            result([
                "encrypted": encrypted,
                "tag": tag,
                "nonce": nonce,
            ])
        } catch CryptoError.runtimeError(let errorMessage) {
            result(FlutterError(code: "InvalidArgumentError", message: errorMessage.localizedLowercase, details: nil))
        } catch {
            result(FlutterError(code: "EncryptionError", message:"Failed to encrypt string: " + error.localizedDescription, details: nil))
        }
    }

    private func decrypt(args: [String: Any], result: @escaping FlutterResult) throws {
        do {
            guard let cipherData = (args["encrypted"] as? FlutterStandardTypedData)?.data else {
                result(parameterError(name: "encrypted"))
                return
            }
            guard let key = (args["key"] as? FlutterStandardTypedData)?.data else {
                result(parameterError(name: "key"))
                return
            }
            guard let nonce = (args["nonce"] as? FlutterStandardTypedData)?.data else {
                result(parameterError(name: "nonce"))
                return
            }
            guard let tag = (args["tag"] as? FlutterStandardTypedData)?.data else {
                result(parameterError(name: "tag"))
                return
            }

            let decryptedData = try decryptData(cipherData: cipherData, keyData: key, nonceData: nonce, tagData: tag)
            
            result(FlutterStandardTypedData(bytes: decryptedData))
            
        } catch CryptoError.runtimeError(let errorMessage) {
            result(FlutterError(code: "InvalidArgumentError", message: errorMessage.localizedLowercase, details: nil))
        } catch {
            result(FlutterError(code: "EncryptionError", message:"Failed to decrypt: " + error.localizedDescription, details: nil))
        }
    }
    
    private func decryptString(args: [String: Any], result: @escaping FlutterResult) throws {
        do {
            guard let inputEncoding = args["inputEncoding"] as? String else {
                result(parameterError(name: "inputEncoding"))
                return
            }
            
            if (inputEncoding != "base64" && inputEncoding != "hex") {
                result(FlutterError(code: "Input encoding eror", message: "Input key encoding should be in 'base64' or 'hex'", details: nil))
                return
            }
        
            guard let encryptedString = args["encryptedString"] as? String else {
                result(parameterError(name: "encryptedString"))
                return
            }
            guard let key = args["key"] as? String else {
                result(parameterError(name: "key"))
                return
            }
            guard let nonce = args["nonce"] as? String else {
                result(parameterError(name: "nonce"))
                return
            }
            guard let tag = args["tag"] as? String else {
                result(parameterError(name: "tag"))
                return
            }
            
            let cipherData = inputEncoding == "base64" ? Data(base64Encoded: encryptedString)! : Data(hexString: encryptedString)!
            let keyData = inputEncoding == "base64" ? Data(base64Encoded: key)! : Data(hexString: key)!
            let nonceData = inputEncoding == "base64" ? Data(base64Encoded: nonce)! : Data(hexString: nonce)!
            let tagData = inputEncoding == "base64" ? Data(base64Encoded: tag)! : Data(hexString: tag)!

            let decryptedData = try decryptData(cipherData: cipherData, keyData: keyData, nonceData: nonceData, tagData: tagData)
            
            result(String(decoding: decryptedData, as: UTF8.self))
            
        } catch CryptoError.runtimeError(let errorMessage) {
            result(FlutterError(code: "InvalidArgumentError", message: errorMessage.localizedLowercase, details: nil))
        } catch {
            result(FlutterError(code: "EncryptionError", message:"Failed to decrypt: " + error.localizedDescription, details: nil))
        }
    }
    
    private func parameterError(name: String) -> FlutterError {
        return FlutterError(code: "INVALID_ARGUMENT", message: "Parameter '\(name)' is missing or invalid", details: nil)
    }
}

File: android/src/main/kotlin/com/hayr_hotoca/flutter_chaha20_poly1305/FlutterChacha20Poly1305Plugin.kt
package com.hayr_hotoca.flutter_chaha20_poly1305

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.google.crypto.tink.subtle.ChaCha20Poly1305
import org.spongycastle.util.encoders.Base64
import org.spongycastle.util.encoders.Hex

class Output(val nonce: ByteArray, val tag: ByteArray, val encrypted: ByteArray)

/** FlutterChacha20Poly1305Plugin */
class FlutterChacha20Poly1305Plugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_chacha20_poly1305")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "encrypt") {
      encrypt(call, result)
    } else if (call.method == "encryptString") {
      encryptString(call, result)
    } else if (call.method == "decrypt") {
      decrypt(call, result)
    } else if (call.method == "decryptString") {
      decryptString(call, result)
    } else {
      result.notImplemented()
    }
  }

  private fun encrypt(call: MethodCall, result: Result) {
    try {
      val data = call.argument<ByteArray>("data")!!
      val key = call.argument<ByteArray>("key")!!

      val encryptedData = encryptData(data, key)

      result.success(
              hashMapOf(
                      "encrypted" to encryptedData.encrypted,
                      "nonce" to encryptedData.nonce,
                      "tag" to encryptedData.tag
              )
      )
    } catch (e: Exception) {
      result.error("EncryptionError", "Failed to encrypt: " + e.localizedMessage, e)
    }
  }

  private fun encryptString(call: MethodCall, result: Result) {
    try {
      val keyEncoding = call.argument<String>("keyEncoding")
      val outputEncoding = call.argument<String>("outputEncoding")
      if (keyEncoding != "base64" && keyEncoding != "hex")
        return result.error("Input encoding eror", "Input key encoding should be in 'base64' or 'hex'", null)
      if (outputEncoding != "base64" && outputEncoding != "hex")
        return result.error("Output encoding eror", "Output encoding should be in 'base64' or 'hex'", null)

      val key = call.argument<String>("key")!!
      val keyData = if (keyEncoding == "base64") Base64.decode(key) else  Hex.decode(key)

      val string = call.argument<String>("string")!!
      val plainData = string.toByteArray()

      val encryptedData = encryptData(plainData, keyData)

      result.success(
              hashMapOf(
                      "encrypted" to if (outputEncoding == "base64") Base64.toBase64String(encryptedData.encrypted) else Hex.toHexString(encryptedData.encrypted),
                      "nonce" to if (outputEncoding == "base64") Base64.toBase64String(encryptedData.nonce) else Hex.toHexString(encryptedData.nonce),
                      "tag" to if (outputEncoding == "base64") Base64.toBase64String(encryptedData.tag) else Hex.toHexString(encryptedData.tag),
              )
      )
    } catch (e: Exception) {
      result.error("EncryptionError", "Failed to encrypt string: " + e.localizedMessage, e)
    }
  }

  private fun decrypt(call: MethodCall, result: Result) {
    try {
      val encrypted = call.argument<ByteArray>("encrypted")!!
      val key = call.argument<ByteArray>("key")!!
      val nonce = call.argument<ByteArray>("nonce")!!
      val tag = call.argument<ByteArray>("tag")!!
      val cipherData = nonce + encrypted + tag

      val decrypted = decryptData(cipherData, key)

      result.success(decrypted)
    } catch  (e: Exception) {
      result.error("EncryptionError", "Failed to decrypt: " + e.localizedMessage, e)
    }
  }

  private fun decryptString(call: MethodCall, result: Result) {
    try {
      val inputEncoding = call.argument<String>("inputEncoding")
      if (inputEncoding != "base64" && inputEncoding != "hex") return result.error("Input encoding eror", "Input encoding should be in 'base64' or 'hex'", null)

      val encryptedString = call.argument<String>("encryptedString")!!
      val key = call.argument<String>("key")!!
      val keyData = if (inputEncoding == "base64") Base64.decode(key) else Hex.decode(key)
      val nonce = call.argument<String>("nonce")!!
      val tag = call.argument<String>("tag")!!
      val cipherData = if (inputEncoding == "base64")
        Base64.decode(nonce) + Base64.decode(encryptedString) + Base64.decode(tag)
      else
        Hex.decode(nonce + encryptedString + tag)

      val decrypted = decryptData(cipherData, keyData)

      result.success(decrypted.toString(Charsets.UTF_8))
    } catch  (e: Exception) {
      result.error("EncryptionError", "Failed to decrypt string: " + e.localizedMessage, e)
    }
  }

  private fun encryptData(plainData: ByteArray, keyData: ByteArray): Output {
    val cipher = ChaCha20Poly1305(keyData)
    val sealed = cipher.encrypt(plainData, null)
    val ivData = sealed.sliceArray(0..11)
    val tagData = sealed.sliceArray(sealed.size-16..sealed.size-1)
    val encryptedData = sealed.sliceArray(12..sealed.size-17)

    return Output(ivData, tagData, encryptedData)
  }

  private fun decryptData(cipherData: ByteArray, key: ByteArray): ByteArray {
    val cipher = ChaCha20Poly1305(key)
    val decrypted = cipher.decrypt(cipherData, null)

    return decrypted
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}
